### PR TITLE
Renames the service name to include the app name

### LINKED ISSUE #4550: `odo list --app appName --project projectName` confusing components with the same name in different applications
/kind bug

## What versions of software are you using?

**Operating System:** Linux

**Output of `odo version`:** 2.0.7

## How did you run odo exactly?

Create nodejs devfile component in demo application

```
$ odo create nodejs nodejs  --context /home/denis/p/test-ws/comp1 --app app --project project-name --starter=nodejs-starter
Devfile Object Validation
 •  Checking devfile existence  ...

 ✓  Checking devfile existence [16368ns]
 •  Creating a devfile component from registry: DefaultDevfileRegistry  ...

 ✓  Creating a devfile component from registry: DefaultDevfileRegistry [20154ns]
Validation
 •  Validating if devfile name is correct  ...

 ✓  Validating if devfile name is correct [16569ns]

Starter Project
 •  Downloading starter project nodejs-starter from https://github.com/odo-devfiles/nodejs-ex.git  ...

 ✓  Downloading starter project nodejs-starter from https://github.com/odo-devfiles/nodejs-ex.git [567ms]

Please use `odo push` command to create the component with source deployed
```
```
$ odo push
Validation
 ✓  Validating the devfile [41140ns]

Creating Kubernetes resources for component nodejs
 ✓  Waiting for component to start [6s]
 ✓  Waiting for component to start [15ms]

Applying URL changes
 ✓  URL http-3000: http://http-3000-nodejs-project-name.apps-crc.testing/ created

Syncing to component nodejs
 ✓  Checking files for pushing [1ms]
 ✓  Syncing files to the component [527ms]

Executing devfile commands for component nodejs
 ✓  Waiting for component to start [4ms]
 ✓  Executing install command "npm install" [4s]
 ✓  Executing run command "npm start" [1s]

Pushing devfile component nodejs
 ✓  Changes successfully pushed to component
```
```
$ odo list --app app --project project-name
Devfile Components: 
APP     NAME       PROJECT          TYPE       STATE
app     nodejs     project-name     nodejs     Pushed
```
```
$ odo create nodejs nodejs  --context /home/denis/p/test-ws/comp2 --app app1 --project project-name --starter=nodejs-starter
Devfile Object Validation
 •  Checking devfile existence  ...

 ✓  Checking devfile existence [15503ns]
 •  Creating a devfile component from registry: DefaultDevfileRegistry  ...

 ✓  Creating a devfile component from registry: DefaultDevfileRegistry [21721ns]
Validation
 •  Validating if devfile name is correct  ...

 ✓  Validating if devfile name is correct [31685ns]

Starter Project
 •  Downloading starter project nodejs-starter from https://github.com/odo-devfiles/nodejs-ex.git  ...

 ✓  Downloading starter project nodejs-starter from https://github.com/odo-devfiles/nodejs-ex.git [517ms]

Please use `odo push` command to create the component with source deployed
```
```
odo" push
 ✓  Waiting for component to start [8ms]

Validation
 ✓  Validating the devfile [54309ns]

Creating Kubernetes resources for component nodejs
 ✓  Waiting for component to start [14s]
 ✓  Waiting for component to start [23ms]

Applying URL changes
 ✓  URLs are synced with the cluster, no changes are required.

Syncing to component nodejs
 ✓  Checking file changes for pushing [899428ns]
 ✓  Syncing files to the component [683ms]

Executing devfile commands for component nodejs
 ✓  Waiting for component to start [3ms]
 ✓  Executing install command "npm install" [5s]
 ✓  Executing run command "npm start" [1s]

Pushing devfile component nodejs
 ✓  Changes successfully pushed to component
```

## Actual behavior

odo list does not see app/nodejs component anymore

```
odo list --app app --project project-name
There are no components deployed.
```
But list app1/nodejs without issues

```
odo list --app app1 --project project-name
Devfile Components: 
APP      NAME       PROJECT          TYPE       STATE
app1     nodejs     project-name     nodejs     Pushed
```
Running odo delete for component in app1 below

```
odo delete nodejs -f --app app1 --project project-name --all 
Gathering information for component nodejs
 •  Checking status for component  ...

 ✓  Checking status for component [22ms]

Deleting component nodejs
 •  Deleting Kubernetes resources for component  ...

 ✓  Deleting Kubernetes resources for component [22ms]
 ✓  Successfully deleted component

Deleting local config
 ✓  Successfully deleted env file
 ✓  Successfully deleted devfile.yaml file
```
leads to deletion both app/nodejs and app1/nodejs components

```
odo list --app app --project project-name
There are no components deployed.
odo list --app app --project project-name
There are no components deployed.
```

## Expected behavior

```
odo list --app app1 --project project-name
Devfile Components: 
APP      NAME       PROJECT          TYPE       STATE
app1     nodejs     project-name     nodejs     Pushed
```
```
odo list --app app --project project-name
Devfile Components: 
APP      NAME       PROJECT          TYPE       STATE
app     nodejs     project-name     nodejs     Pushed

```
```
odo delete nodejs -f --app app1 --project project-name --all 
Gathering information for component nodejs
 •  Checking status for component  ...

 ✓  Checking status for component [22ms]

Deleting component nodejs
 •  Deleting Kubernetes resources for component  ...

 ✓  Deleting Kubernetes resources for component [22ms]
 ✓  Successfully deleted component

Deleting local config
 ✓  Successfully deleted env file
 ✓  Successfully deleted devfile.yaml file
```
```
odo list --app app1 --project project-name
There are no components deployed.
```
```
odo list --app app --project project-name
Devfile Components: 
APP      NAME       PROJECT          TYPE       STATE
app     nodejs     project-name     nodejs     Pushed
```

### PR DESCRIPTION
**Pull request by mik-dass**
**What type of PR is this?**

/kind bug
/kind design

**What does this PR do / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #4550 

**PR acceptance criteria**:

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

- [ ] Update changelog

- [ ] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

Create two components with the same name in different apps
- `odo create nodejs <component-name> --app app --context <directory of the first component>`
- `odo create springboot <component-name> --app app1 --context <directory of the second component>`
- Trigger `odo push` for the both the components
- `oc get service` should list both the services
- `oc get deployments` should list both the deployments
- Check the list output for `odo list --app app` and `odo list --app app1`.

### CODE REVIEW

**Comment by valaparthvi** [comment]
@mik-dass
I followed the steps from issue #4550 
1. odo create nodejs nodejs --starter=nodejs-starter --project=myproject --app app
2. odo push
3. odo create nodejs nodejs --starter=nodejs-starter --project=myproject --app app1
4. odo push
But it failed at step 4.
```
odo push

Validation
 ✓  Validating the devfile [163115ns]

Creating Kubernetes resources for component nodejs
 ✓  Waiting for component to start [4s]
 ✓  Waiting for component to start [377ms]

Applying URL changes
 ✗  Failed To Update Config To Component Deployed.
Error: url named "http-3000" already exists in the same app named "app1"
```
Is this expected?

I also checked `kubectl get services` which showed services name containing app name, which I think serves the purpose of this PR.
```
kubectl get services
NAME                          TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)             AGE
nodejs-app                    ClusterIP   172.30.110.139   <none>        3000/TCP            2m23s
nodejs-app1                   ClusterIP   172.30.93.13     <none>        3000/TCP            62s
postgresql-operator-metrics   ClusterIP   172.30.188.67    <none>        8383/TCP,8686/TCP   2d1h
```

Is there something else I should be looking at/testing?

**Comment by mik-dass** [comment]
> Applying URL changes
>  ✗  Failed To Update Config To Component Deployed.
> Error: url named "http-3000" already exists in the same app named "app1"
> 
> Is this expected?

This is a bug as both the routes try to use the same name `http-3000`. I will raise a PR for the fix.

> Is there something else I should be looking at/testing?

The list output of `odo list --app app` and `odo list --app app1`

**Comment by valaparthvi** [comment]
Based on https://github.com/openshift/odo/pull/4858#issuecomment-872893125 and output of odo list --app app and odo list --app app, the functionality lgtm.

/approve

File: pkg/kclient/errors.go [hunk 1]
```diff
@@ -19,3 +19,12 @@ type DeploymentNotFoundError struct {
...
 	return fmt.Sprintf("deployment not found for the selector: %s", e.Selector)
 }
+
+// ServiceNotFoundError returns an error if no service is found with the selector
+type ServiceNotFoundError struct {
+	Selector string
+}
+
+func (e *ServiceNotFoundError) Error() string {
+	return fmt.Sprintf("service not found for the selector: %s", e.Selector)
```
Comment (feloy) [inline pkg/kclient/errors.go:29]: Could you use `service not found for the selector %q` instead? (to reserve `...: ...` for wrapped error)

File: pkg/kclient/services.go [hunk 2]
```diff
@@ -47,3 +49,35 @@ func (c *Client) ListServices(selector string) ([]corev1.Service, error) {
...
+	return nil
+}
+
+// GetOneService retrieves the service with the given component and app name
+func (c *Client) GetOneService(componentName, appName string) (*corev1.Service, error) {
+	return c.GetOneServiceFromSelector(componentlabels.GetSelector(componentName, appName))
+}
+
+// GetOneServiceFromSelector returns the service object associated with the given selector.
+// An error is thrown when exactly one service is not found for the selector.
```
Comment (feloy) [inline pkg/kclient/services.go:68]: The second line of this comment would also be useful on `GetOneService`

File: pkg/kclient/services.go [hunk 3]
```diff
@@ -47,3 +49,35 @@ func (c *Client) ListServices(selector string) ([]corev1.Service, error) {
...
+func (c *Client) GetOneServiceFromSelector(selector string) (*corev1.Service, error) {
+	services, err := c.ListServices(selector)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get Deployments for the selector: %v", selector)
```
Comment (feloy) [inline pkg/kclient/services.go:72]: As discussed with @mohammedzee1000 on the chat, it is recommended to use fmt.Errorf("unable to get Deployments for the selector %q: %w", selector, err)

Reference: https://blog.golang.org/go1.13-errors

**Comment by mik-dass** [comment]
```
          <*errors.errorString | 0xc000ac4180>: {
              s: "cmd [ps -ef] failed with error Signal 23 (URG) caught by ps (3.3.15).\nps:ps/display.c:66: please report this bug\ncommand terminated with exit code 1\n on pod feduwc-app-c8f869b9f-xg7dg",
          }
          cmd [ps -ef] failed with error Signal 23 (URG) caught by ps (3.3.15).
          ps:ps/display.c:66: please report this bug
```
/retest

**Comment by feloy** [comment]
> ```
>           <*errors.errorString | 0xc000ac4180>: {
>               s: "cmd [ps -ef] failed with error Signal 23 (URG) caught by ps (3.3.15).\nps:ps/display.c:66: please report this bug\ncommand terminated with exit code 1\n on pod feduwc-app-c8f869b9f-xg7dg",
>           }
>           cmd [ps -ef] failed with error Signal 23 (URG) caught by ps (3.3.15).
>           ps:ps/display.c:66: please report this bug
> ```
> 
> /retest

I would be curious to know which process is sending a URG signal to ps, and why. oc? @kadel any clue?

**Pull request merged**